controller/src/traffic_split.rs

```diff
@@ -166,17 +173,44 @@ async fn patch(
     client: kube::Client,
     params: &PatchParams,
     timeout: time::Duration,
-    FailoverUpdate { target, backends }: FailoverUpdate,
+    FailoverUpdate {
+        target,
+        backends,
+        primary_active,
+    }: FailoverUpdate,
 ) {
-    let namespace = target.namespace.expect("namespace must be set");
-    let api = Api::<TrafficSplit>::namespaced(client, &namespace);
-    let name = target.name;
+    let namespace = target.namespace.as_ref().expect("namespace must be set");
+    let api = Api::<TrafficSplit>::namespaced(client.clone(), namespace);
+    let name = &target.name;
     tracing::debug!("patching trafficsplit");
 
-    let patch = mk_patch(&name, &backends);
+    let event_reporter = events::Reporter {
+        controller: "linkerd-failover".to_string(),
+        instance: None,
+    };
+    let event_recorder = events::Recorder::new(client, event_reporter, target.clone().into());
+    let description = if primary_active {
+        format!("trafficsplit/{} switching traffic to primary", name)
+    } else {
+        format!("trafficsplit/{} failing over to fallbacks", name)
+    };
+    if let Err(error) = event_recorder
+        .publish(events::Event {
+            type_: events::EventType::Normal,
+            reason: FAILOVER.to_string(),
+            note: Some(description),
+            action: FAILOVER.to_string(),
+            secondary: None,
+        })
+        .await
+    {
+        tracing::error!(%error, "failed to record event");
+    }
```

Comment:
perhaps worth being in its own `async fn`? kind of decoupled from the patching logic...

Also, would it be better to emit this event after patching succeeds so that we do't have the possibility of false-positives?